File: projecteuler/three.erl
#!/usr/bin/env escript
%%%-------------------------------------------------------------------
%%% @author zhuwei
%%% @copyright (C) 2015, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 29. 七月 2015 下午8:01
%%%-------------------------------------------------------------------
-module(three).
-author("zhuwei").
-mode(compile).

%% API
-export([main/1]).

helper(N) ->
  SqrtN = round(math:sqrt(N)),
  MaxFactor = 1,
  InitFactor = 2,
  max_factor_prime(InitFactor, SqrtN, N, MaxFactor).

max_factor_prime(_InitFactor, 0, _N, MaxFactor) ->
  MaxFactor;
max_factor_prime(InitFactor, SqrtN, N, MaxFactor) ->
  if (N rem InitFactor =:= 0) ->
    if (MaxFactor < InitFactor) ->
      max_factor_prime(InitFactor + 1, SqrtN-1, N div InitFactor, InitFactor);
      'true' -> max_factor_prime(InitFactor + 1, SqrtN-1, N div InitFactor, MaxFactor)
    end;
    'true' -> max_factor_prime(InitFactor+1, SqrtN-1, N, MaxFactor)
  end.

main(_) ->
  io:format("~p~n", [helper(600851475143)]),
  'ok'.

File: projecteuler/three.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = 'zhuwei'

import math


def is_prime(numbers):
    if numbers <= 1:
        return False
    sqrt_number = int(math.sqrt(numbers)) + 1
    for i in xrange(2, sqrt_number):
        if numbers % i == 0:
            return False

    return True


def max_prime_factor(numbers):
    sqrt_n = int(math.sqrt(numbers))
    max_factor = 1
    for i in xrange(2, sqrt_n):
        if numbers % i == 0:
            if i > max_factor:
                max_factor = i
            numbers /= i
    return max_factor

if __name__ == '__main__':
    # number = 13195
    number = 600851475143
    print "max prime factor %s" % max_prime_factor(number)